### PR TITLE
Draft: 1.18.2 Port

### PR DESCRIPTION
**Pull request by morguldir**
Issues from #4682 still apply, and there are some extra bugs as well, but probably ready for a beta

Most of the changes are adding holders where required, using `DeferredRegistry` for everything (items, blocks, recipes, etc.), changing `save` to `saveAdditional`, changes to structure generation, and more fun forge changes

Thanks @McJty for the tutorials :D

### This port is still not official, so make sure you trust me before downloading the jar
https://github.com/morguldir/Ice_and_Fire/releases/tag/1.18.2-latest-snapshot

### CODE REVIEW

File: src/main/java/com/github/alexthe666/iceandfire/entity/EntityMyrmexQueen.java [hunk 1]
```diff
@@ -372,4 +372,9 @@ public int getVillagerXp() {
     public boolean showProgressBar() {
         return false;
     }
+
+    @Override
+    public boolean isClientSide() {
+        return false;
```
Comment (TheBv) [inline src/main/java/com/github/alexthe666/iceandfire/entity/EntityMyrmexQueen.java:378]: Everything else return this.getLevel().isClientSide;

Comment (TheBv) [inline src/main/java/com/github/alexthe666/iceandfire/entity/EntityMyrmexQueen.java:378]: Everything else returns this.getLevel().isClientSide;

File: src/main/java/com/github/alexthe666/iceandfire/world/gen/mixin/NoLakesInStructuresMixin.java [hunk 2]
```diff
@@ -17,17 +20,23 @@
...
-    )
-    private void iaf_noLakesInMausoleum(FeaturePlaceContext<BlockStateConfiguration> context, CallbackInfoReturnable<Boolean> cir) {
-        SectionPos sectionPos = SectionPos.of(context.origin());
-        WorldGenLevel level = context.level();
-        StructureFeature<NoneFeatureConfiguration> structure = IafWorldRegistry.MAUSOLEUM.get();
-        if (level.startsForFeature(sectionPos, structure).findAny().isPresent()) {
-            cir.setReturnValue(false);
-        }
-    }
+    // TODO: morguldir: Revisit this later
```
Comment (TheBv) [inline src/main/java/com/github/alexthe666/iceandfire/world/gen/mixin/NoLakesInStructuresMixin.java]: Check out https://github.com/TelepathicGrunt/RepurposedStructures/blob/1.18-Forge/src/main/java/com/telepathicgrunt/repurposedstructures/mixin/features/NoLakesInStructuresMixin.java for this

File: src/main/java/com/github/alexthe666/iceandfire/item/ItemSeaSerpentScales.java [hunk 3]
```diff
@@ -16,7 +16,7 @@ public class ItemSeaSerpentScales extends ItemGeneric {
     private final String colorName;
 
     public ItemSeaSerpentScales(String colorName, ChatFormatting color) {
-        super("sea_serpent_scales_" + colorName);
+        //super("sea_serpent_scales_" + colorName);
```
Comment (TheBv) [inline src/main/java/com/github/alexthe666/iceandfire/item/ItemSeaSerpentScales.java]: nit: remove comment

File: src/main/java/com/github/alexthe666/iceandfire/datagen/IafBiomeTagsProvider.java [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+import javax.annotation.Nullable;
+import java.util.Map;
+
+public class IafBiomeTagsProvider extends BiomeTagsProvider {
+    public IafBiomeTagsProvider(DataGenerator generator, String modId, @Nullable ExistingFileHelper existingFileHelper) {
+        super(generator, modId, existingFileHelper);
+    }
+
+    @Override
+    protected void addTags() {
```
Comment (TheBv) [inline src/main/java/com/github/alexthe666/iceandfire/datagen/IafBiomeTagsProvider.java:26]: Does this mean that minecraft now handles in which biomes structures can spawn on it's own?

**Comment by TheBv** [comment]
Looks good. I've yet to test the version myself but thanks for all your work :)

File: src/main/java/com/github/alexthe666/iceandfire/datagen/IafBiomeTagsProvider.java (hunk 4 above)
Comment (morguldir) [reply src/main/java/com/github/alexthe666/iceandfire/datagen/IafBiomeTagsProvider.java:26]: Yep!, heres a fun quote from @McJty :D :
> This code, to port this, the easiest way to port this, is to just delete it, we don't need it anymore

https://www.youtube.com/watch?v=Bw1m4guXMmQ

File: build.gradle [hunk 5]
```diff
@@ -46,11 +54,11 @@ if (build_id != null) {
 group = 'com.github.alexthe667' // http://maven.apache.org/guides/mini/guide-naming-conventions.html
 archivesBaseName = 'iceandfire'
 
-java.toolchain.languageVersion = JavaLanguageVersion.of(16)
+java.toolchain.languageVersion = JavaLanguageVersion.of(17)
 
 println "Java: ${System.getProperty 'java.version'}, JVM: ${System.getProperty 'java.vm.version'} (${System.getProperty 'java.vendor'}), Arch: ${System.getProperty 'os.arch'}"
 minecraft {
-    mappings channel: 'official', version: '1.17.1'
+    mappings channel: 'parchment', version: '2022.11.06-1.18.2'
```
Comment (morguldir) [inline build.gradle:61]: @TheBv thoughts?

File: src/main/java/com/github/alexthe666/iceandfire/datagen/IafBiomeTagsProvider.java (hunk 4 above)
Comment (TheBv) [reply src/main/java/com/github/alexthe666/iceandfire/datagen/IafBiomeTagsProvider.java:26]: Okay. I'll disable the biome config for those structures then.

**Pull request merged**